tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Shared/SpanPointers.cs

```diff
@@ -0,0 +1,95 @@
+// <copyright file="SpanPointers.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.Shared
+{
+    /// <summary>
+    ///  SpanPointer helper methods
+    /// </summary>
+    internal static class SpanPointers
+    {
+        // The pointer direction will always be down. The serverless agent handles cases where the
+        // direction is up.
+        private const string DownDirection = "d";
+        private const string LinkKind = "span-pointer";
+        private const int SpanPointerHashSizeBytes = 16;
+        private const string S3PtrKind = "aws.s3.object";
+
+        // S3 hashing rules: https://github.com/DataDog/dd-span-pointer-rules/blob/main/AWS/S3/Object/README.md
+        public static void AddS3SpanPointer(Span span, string bucketName, string key, string eTag)
+        {
+            var components = new[] { bucketName, key, StripQuotes(eTag) };
+            var hash = GeneratePointerHash(components);
+            var spanLinkAttributes = new List<KeyValuePair<string, string>>
```

Comment:
Won't make much different in this case, but in general it's good practice to provide the size of the list (if you know it) to avoid extra work/allocations as the list grows.
```suggestion
            var spanLinkAttributes = new List<KeyValuePair<string, string>>(4)
```